control-plane/p4infoApi.h

```diff
@@ -0,0 +1,185 @@
+#ifndef CONTROL_PLANE_P4INFOAPI_H_
+#define CONTROL_PLANE_P4INFOAPI_H_
+
+#include <optional>
+
+#include "control-plane/p4RuntimeArchHandler.h"
+#include "control-plane/p4RuntimeSerializer.h"
+#include "lib/cstring.h"
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#pragma GCC diagnostic ignored "-Wpedantic"
+#include "p4/config/v1/p4info.pb.h"
+#pragma GCC diagnostic pop
+
+namespace P4::ControlPlaneAPI {
+
+/// Generic meta function which searches an object by @name in the given range
+/// and @returns the P4Runtime representation, or std::nullopt if none is found.
+/// TODO: Should this return an optional const reference? The advantage is that the value needs to
+/// be explicitly unpacked.
```

Comment:
I was thinking of something like an `optional_ref` type. Google does have it, but it is not necessarily widely available: 
https://chromium.googlesource.com/chromium/src/base/+/refs/heads/main/types/optional_ref.h#17

It bridges this gap between returning just a pointer and a non-nullable type and makes it very explicit that you either return a reference to an existing value or an error. 

>FWIW: At Google it is understood that pointer dereferences always have to be checked, and we use references instead of pointers when we know an object cannot be null.

I think the checking is what causes issues, at least I have run into many cases where this ambiguity can cause problems. At least on the open-source side where you have many different contributors with different styles.

But I am not going to implement this on this PR ha.

Reply:
I haven't seen `optional_ref` before, that looks nice!

> At least on the open-source side where you have many different contributors with different styles.

Yes, understood that conventions are a lot harder to apply consistently in open source, hence the "FWIW" prefix :)